### ISSUE TITLE
Hang during shutdown in Microsoft.VisualStudio.LanguageServices.Remote.ServiceHubRemoteHostClient.UnregisterGlobalOperationNotifications

### ISSUE THREAD

**Issue by jasonmalinowski** [dotnet/roslyn#40964]
**Version Used**: 16.5.29707.192 built by: MASTER

This is a random hang I hit during shutdown. The UI was already closed I think but it seems we deadlocked somehow shutting down the OOP process:

```
 	mscorlib.dll!System.Threading.Monitor.Wait(object obj, int millisecondsTimeout, bool exitContext) Line 203	C#
 	mscorlib.dll!System.Threading.Monitor.Wait(object obj, int millisecondsTimeout) Line 213	C#
 	mscorlib.dll!System.Threading.ManualResetEventSlim.Wait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 688	C#
 	mscorlib.dll!System.Threading.Tasks.Task.SpinThenBlockingWait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 3320	C#
 	mscorlib.dll!System.Threading.Tasks.Task.InternalWait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 3259	C#
 	mscorlib.dll!System.Threading.Tasks.Task.Wait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 3167	C#
 	[Waiting on Async Operation, Open Tasks View in Parallel Stacks Window to view Async Call Stacks]	
>	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Remote.ServiceHubRemoteHostClient.UnregisterGlobalOperationNotifications() Line 213	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Remote.ServiceHubRemoteHostClient.OnStopped() Line 165	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Remote.RemoteHostClientServiceFactory.RemoteHostClientService.Disable() Line 141	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService.AbstractPackage<System.__Canon, System.__Canon>.DisableRemoteHostClientService() Line 201	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService.AbstractPackage<Microsoft.VisualStudio.LanguageServices.CSharp.LanguageService.CSharpPackage, Microsoft.VisualStudio.LanguageServices.CSharp.LanguageService.CSharpLanguageService>.Dispose.AnonymousMethod__18_0() Line 141	C#
```

**Comment by jasonmalinowski** [comment]
Dumps of devenv and of our OOP processes are available at \\mlangfs1\public\jasonmal\Dumps\Roslyn\40964

**Comment by jasonmalinowski** [comment]
I also saw this shutting down again, this time with extra work happening on a background thread. This was being called in a loop:

```
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.ExternalAccess.UnitTesting.UnitTestingIncrementalAnalyzer.RemoveDocument(Microsoft.CodeAnalysis.DocumentId documentId) Line 45	C#
>	Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.SolutionCrawler.SolutionCrawlerRegistrationService.WorkCoordinator.IncrementalAnalyzerProcessor.NormalPriorityProcessor.RemoveDocument(System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.SolutionCrawler.IIncrementalAnalyzer> analyzers, Microsoft.CodeAnalysis.DocumentId documentId) Line 461	C#
Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.SolutionCrawler.SolutionCrawlerRegistrationService.WorkCoordinator.IncrementalAnalyzerProcessor.NormalPriorityProcessor.RemoveDocument(Microsoft.CodeAnalysis.DocumentId documentId) Line 457	C#
Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.SolutionCrawler.SolutionCrawlerRegistrationService.WorkCoordinator.IncrementalAnalyzerProcessor.NormalPriorityProcessor.ProcessDocumentAsync(System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.SolutionCrawler.IIncrementalAnalyzer> analyzers, Microsoft.CodeAnalysis.SolutionCrawler.SolutionCrawlerRegistrationService.WorkCoordinator.WorkItem workItem, System.Threading.CancellationToken cancellationToken) Line 362	C#
 	[Resuming Async Method]	
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start<Microsoft.CodeAnalysis.SolutionCrawler.SolutionCrawlerRegistrationService.WorkCoordinator.IncrementalAnalyzerProcessor.NormalPriorityProcessor.<ProcessDocumentAsync>d__32>(ref Microsoft.CodeAnalysis.SolutionCrawler.SolutionCrawlerRegistrationService.WorkCoordinator.IncrementalAnalyzerProcessor.NormalPriorityProcessor.<ProcessDocumentAsync>d__32 stateMachine)	Unknown
Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.SolutionCrawler.SolutionCrawlerRegistrationService.WorkCoordinator.IncrementalAnalyzerProcessor.NormalPriorityProcessor.TryProcessOneHigherPriorityDocumentAsync() Line 282	C#
Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.SolutionCrawler.SolutionCrawlerRegistrationService.WorkCoordinator.IncrementalAnalyzerProcessor.NormalPriorityProcessor.ExecuteAsync() Line 142	C#
Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.SolutionCrawler.IdleProcessor.ProcessAsync() Line 104	C#
```

@vatsalyaagrawal @jinujoseph This seems to have some reliability and dogfood impact, can we get somebody to look at it?

**Comment by CyrusNajmabadi** [comment]
@mavasani @shyamnamboodiripad This appears to be solution crawler + unit testing.

1. @mavasani  during shutdown should we just cancel all work and immediately return ourselves?  Are we already cancelling?
2. @shyamnamboodiripad We may already be cancelling.  We recently changed  RemoveDocument and RemoveProject to have async overloads.  Could you update your integration pieces to use that pattern so that you can be effectively stopped during times like shutdown?

**Comment by shyamnamboodiripad** [comment]
We don't do anything on RemoveProject. Yes we could move to a cancellable RemoveDocument - but it may be better to tackle this at the same time as we move our analysis to the OOP crawler. It sounds like the problem may not be purely cancellation though as Jason mentioned that RemoveDocument was being called in a loop (presumably this happens for every document in the solution on close). Couldn't you stop calling RemoveDocument() for subsequent documents if you know VS is being shutdown? If VS is not being closed, I suppose you can't cancel before all removals are processed anyways...

**Comment by CyrusNajmabadi** [comment]
> Couldn't you stop calling RemoveDocument() for subsequent documents if you know VS is being shutdown?

**Comment by tmat** [comment]
The original issue was fixed by https://github.com/dotnet/roslyn/pull/43752.

Should we file another issue for the unit testing?

**Comment by CyrusNajmabadi** [comment]
Fixed with https://github.com/dotnet/roslyn/pull/43752
